tests/test_aadt_mapping.py

```diff
@@ -0,0 +1,207 @@
+from hwtypes import SMTBitVector, SMTBit
+from peak.assembler import Assembler
+from peak.assembler import AssembledADT
+from peak.mapper.utils import Tag, Match, generic_aadt_smt
+from examples.min_pe.sim import gen_sim
+from examples.min_pe.sim import gen_isa
+from functools import reduce
+import pysmt.shortcuts as smt
+from pysmt.logics import BV
+import typing as tp
+import math
+
+SBV = SMTBitVector
+
+def add(a, b):  return a + b
+def sub(a, b):  return b - a
+def and_(a, b): return a & b
+def nand(a, b): return ~(a & b)
+def or_(a, b):  return a | b
+def nor(a, b):  return ~(b | a)
+def mul(a, b):  return a * b
+def shftr(a, b): return a >> b
+def shftl(a, b): return a << b
+targets = (
+    add,
+    sub,
+    and_,
+    nand,
+    or_,
+    nor,
+    mul,
+    shftr,
+    shftl,
+)
+Word, Bit, Inst, sim = gen_sim(SBV.get_family())
+
+T = Tuple[Word, Bit]
+S = Sum[Word, T]
+
+Inst_aadt = AssembledADT[Inst, Assembler, SBV]
+
+
+forms = [
+    {("operand_1",):Word},
+    {("operand_1",):T}
+]
+
+
+#Forms:
+
+class Unbound: pass
+
+#Manually creating all the forms
+#forms:
+# operand_1 -> Word
+# opeanrd_1 -> Tuple
+forms = []
+
+#form 0
+forms.append([
+    (("operand_1",), Word),
+])
+
+#form 1
+forms.append([
+    (("operand_1",), T),
+])
+
+#Manually create all the bindings
+form_bindings = [[], []]
+
+#form 0 (operand_1, Word)
+#[a -> operand_0, b -> (operand_1, Word)]
+#[b -> operand_0, a -> (operand_1, Word)]
+
+form_bindings[0].append([
+    (("a",), ("operand_0",)),
+    (("b",), ("operand_1", Word)),
+    (Unbound, ("Opcode",))
+])
+form_bindings[0].append([
+    (("b",), ("operand_0",)),
+    (("a",), ("operand_1", Word)),
+    (Unbound, ("Opcode",))
+])
+
+#form 1: (operand_1, Tuple[Word, Bit])
+#[a -> operand_0, b -> (operand_1, Tuple, 0)]
+#[b -> operand_0, a -> (operand_1, Tuple, 0)]
+form_bindings[1].append([
+    (("a",), ("operand_0",)),
+    (("b",), ("operand_1", T, 0)),
+    (Unbound, ("operand_1", T, 1)),
+    (Unbound, ("Opcode",))
+])
+form_bindings[1].append([
+    (("b",), ("operand_0",)),
+    (("a",), ("operand_1", T, 0)),
+    (Unbound, ("operand_1", T, 1)),
+    (Unbound, ("Opcode",))
+])
+
+arch_inst, arch_varmap = generic_aadt_smt(Inst_aadt)
+
+#Check that manually created forms and bindings are consistent
+for form in forms:
+    for (path, choice) in form:
+        match_path = path + (Match,)
+        assert match_path in arch_varmap
+        assert choice in arch_varmap[match_path]
+for bindings in form_bindings:
+    for binding in bindings:
+        for ir_path, arch_path in binding:
+            assert arch_path in arch_varmap
+
+#Automatically create the form/binding precondition
+num_forms = len(forms)
+max_bindings = max(len(b) for b in form_bindings)
+
+form_SBV = SBV[num_forms]
+binding_SBV = SBV[max_bindings]
+
+form_var = form_SBV()
+binding_var = binding_SBV()
+
+#arbitrary precision accuract log2 function
+#math.log is not accurate for large powers of 2
+def log2(x):
```

Comment:
is x here a positive python int? If so `x.bit_length()-1` works (`-1` to strip two-complement 0).